functions/xkarpenter/main.k

```diff
@@ -0,0 +1,743 @@
+import regex
+import json
+
+import models.io.upbound.platform.aws.v1alpha1 as platformv1alpha1
+import models.io.upbound.aws.iam.v1beta1 as iamv1beta1
+
+import models.io.upbound.aws.cloudwatchevents.v1beta1 as cloudwatcheventsv1beta1
+import models.io.upbound.aws.sqs.v1beta1 as sqsv1beta1
+import models.io.crossplane.kubernetes.v1alpha1 as kubernetesv1alpha1
+
+oxr = platformv1alpha1.XKarpenter {**option("params").oxr}
+# observed composed resources
+_ocds = option("params").ocds
+# desired composite resource
+_dxr = option("params").dxr
+# desired composed resources
+dcds = option("params").dcds
+
+_metadata = lambda name: str -> any {
+    {
+        annotations = {"krm.kcl.dev/composition-resource-name" = name}
+    }
+}
+# Get values from XR
+_id = oxr.spec.parameters.id
+_region = oxr.spec.parameters.region
+_providerConfigName = oxr.spec.parameters.providerConfigName or "default"
+# Generate resource names
+_sqsQueueName = "${_id}-karpenter"
+_healthEventName = "${_id}-healthevent"
+_instanceRebalanceName = "${_id}-instancerebalance"
+_instanceStateChangeName = "${_id}-instancestatechange"
+_spotInterruptName = "${_id}-spotinterrupt"
+
+_statusClusterName = Undefined
```

Comment:
do we need to set this to undefined ? - the default is Undefined or

Reply:
If i recall correctly it fails when it's not mentioned before at all